src/main/daml/Daml/Finance/Holding/Util.daml

```diff
@@ -65,16 +64,18 @@ acquireImpl :
   , HasSignatory t
   ) => t -> Base.Acquire -> Update (ContractId Base.I)
 acquireImpl holding Base.Acquire{newLockers; context; lockType} = do
-  whenSome holding.lock \lock -> do
-    lock.lockers === newLockers
-    lock.lockType === lockType
-    assertMsg "Contract must be locked." $ lock.lockType == Base.Reentrant
-    assertMsg "Contract must not be locked for this context." $ S.notMember context lock.context
-  newLock <- case (holding.lock, lockType) of
-    (Some existingLock, Base.Reentrant) -> pure existingLock with
-      context = context `S.insert` existingLock.context
-    (None, _) -> pure Base.Lock with lockers = newLockers; context = S.singleton context; lockType
-    _ -> abort "Contract is already locked."
+  newLock <- case holding.lock of
+    None -> pure Base.Lock with lockers = newLockers; context = S.singleton context; lockType
+    Some existingLock -> do
+      assertMsg "New lockers must match the existing lockers" $ existingLock.lockers == newLockers
+      assertMsg "New lock type must match the existing lock type" $ existingLock.lockType == lockType
```

Comment:
maybe move the asserts on line 70 and 71 under case `Base.Reentrant`, and change abort message for `Base.Semaphore` to `abort "Semaphore locks can't be locked multiple times"`?